src/lib/transaction_snark/transaction_snark.ml

```diff
@@ -664,12 +711,19 @@ module Base = struct
               ; source_insufficient_balance = false
               ; source_minimum_balance_violation = false
               ; source_bad_timing = false
+              ; fee_payer_balance_update_not_allowed
+              ; source_balance_update_not_allowed = false
+              ; receiver_balance_update_not_allowed = false
+              ; stake_delegation_update_not_allowed
               }
           | Payment ->
               let receiver_account =
                 if Account_id.equal receiver fee_payer then fee_payer_account
                 else receiver_account
               in
+              let receiver_balance_update_not_allowed =
+                not (Account.has_permission ~to_:`Receive receiver_account)
```

Comment:
You'll need to run these check on the account data in the circuit too, otherwise a malicious prover is free to tamper with these values